include/espi.h

```diff
@@ -0,0 +1,769 @@
+ * @param signal the signal to be requested from eSPI master.
+ * @param level the level of signal requested 0b LOW, 1b HIGH.
+ *
+ * @retval -EIO General input / output error, failed request to master.
+ */
+__syscall int espi_receive_vwire(struct device *dev,
+					enum espi_vwire_signal signal,
+					u8_t *level);
+
+static inline int z_impl_espi_receive_vwire(struct device *dev,
+						enum espi_vwire_signal signal,
+						u8_t *level)
+{
+	const struct espi_driver_api *api =
+		(const struct espi_driver_api *)dev->driver_api;
+
+	return api->receive_vwire(dev, signal, level);
+}
+
+/**
+ * @brief Sends SMBus transaction (out-of-band) packet over eSPI bus.
+ *
+ * This routines provides an interface to encapsulate a SMBus transaction
+ * and send into packet over eSPI bus
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param pckt the packet representation of SMBus transaction
+ *
+ * @retval -EIO General input / output error, failed request to master.
+ */
+__syscall int espi_send_oob(struct device *dev, struct espi_oob_packet pckt);
+
+static inline int z_impl_espi_send_oob(struct device *dev,
+					struct espi_oob_packet pckt)
+{
+	const struct espi_driver_api *api =
+		(const struct espi_driver_api *)dev->driver_api;
+
+	return api->send_oob(dev, pckt);
+}
+
+/**
+ * @brief Receives SMBus transaction (out-of-band) packet from eSPI bus.
+ *
+ * This routines provides an interface to receive and decoded a SMBus
+ * transaction from eSPI bus
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param pckt the packet representation of SMBus transaction
+ *
+ * @retval -EIO General input / output error, failed request to master.
+ */
+__syscall int espi_receive_oob(struct device *dev,
+					struct espi_oob_packet pckt);
+
+static inline int z_impl_espi_receive_oob(struct device *dev,
+						struct espi_oob_packet pckt)
+{
+	const struct espi_driver_api *api =
+		(const struct espi_driver_api *)dev->driver_api;
+
+	return api->receive_oob(dev, pckt);
+}
+
+/**
+ * @brief Sends a read request packet for shared flash.
+ *
+ * This routines provides an interface to send a request to read the flash
+ * component shared between the eSPI master and eSPI slaves.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param pckt the representation of read flash transaction
+ *
+ * @retval -ENOTSUP eSPI flash logical channel transactions not supported.
+ * @retval -EBUSY eSPI flash channel is not ready or disabled by master.
+ * @retval -EIO General input / output error, failed request to master.
+ */
+__syscall int espi_read_flash(struct device *dev,
+				struct espi_flash_packet pckt);
+
+static inline int z_impl_espi_flash_read(struct device *dev,
+						struct espi_flash_packet pckt)
+{
+	const struct espi_driver_api *api =
+		(const struct espi_driver_api *)dev->driver_api;
+
+	return api->flash_read(dev, pckt);
```

Comment:
For better consistency, maybe you want to test flash_read here (and return -ENOTSUP accordingly then). As you do it for manage_callback etc...

(either way in fact: either you force each and every driver to implement the functions - and return -ENOTSUP when relevant - or you allow them to let some api function being NULL).


Same on read/write request actually.

Reply:
@tbursztyka  Verified this is addressed in the latest